File: feature-translate/src/main/java/com/andyanika/translator/feature/translate/TranslationFragment.java
package com.andyanika.translator.feature.translate;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import com.andyanika.translator.common.constants.Extras;
import com.andyanika.translator.common.constants.Screens;
import com.andyanika.translator.common.interfaces.ScreenRouter;
import com.andyanika.translator.common.models.ui.DisplayTranslateResult;
import com.jakewharton.rxbinding4.InitialValueObservable;
import com.jakewharton.rxbinding4.widget.RxTextView;

import javax.inject.Inject;

import dagger.android.support.DaggerFragment;

public class TranslationFragment extends DaggerFragment implements TranslationView {
    @Inject
    TranslationPresenter presenter;

    @Inject
    ScreenRouter router;

    private EditText editInput;
    private TextView txtTranslated;
    private View progress;
    private View errorLayout;
    private View clearBtn;
    private View retryBtn;
    private View offlineIcon;
    private Button srcLangBtn;
    private Button dstLangBtn;
    private ImageButton swapLangBtn;
    private InitialValueObservable<CharSequence> textObservable;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_translation, null);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        editInput = view.findViewById(R.id.edit_input);
        textObservable = RxTextView.textChanges(editInput);

        txtTranslated = view.findViewById(R.id.txt_translated);
        progress = view.findViewById(R.id.search_progress);
        errorLayout = view.findViewById(R.id.error_layout);
        retryBtn = view.findViewById(R.id.btn_retry);
        clearBtn = view.findViewById(R.id.btn_clear);
        srcLangBtn = view.findViewById(R.id.btn_lang_src);
        dstLangBtn = view.findViewById(R.id.btn_lang_dst);
        swapLangBtn = view.findViewById(R.id.btn_lang_swap);
        offlineIcon = view.findViewById(R.id.icon_offline);
    }

    @Override
    public void onStart() {
        super.onStart();

        presenter.subscribe(textObservable);
        srcLangBtn.setOnClickListener(v -> router.navigateTo(Screens.SELECT_LANGUAGE, Extras.MODE_SRC));
        dstLangBtn.setOnClickListener(v -> router.navigateTo(Screens.SELECT_LANGUAGE, Extras.MODE_DST));
        swapLangBtn.setOnClickListener(v -> presenter.swapDirection());
        clearBtn.setOnClickListener(v -> presenter.clear());
        retryBtn.setOnClickListener(v -> presenter.translate(editInput.getText().toString()));
    }

    @Override
    public void onStop() {
        presenter.dispose();
        retryBtn.setOnClickListener(null);
        srcLangBtn.setOnClickListener(null);
        dstLangBtn.setOnClickListener(null);

        super.onStop();
    }

    @Override
    public void showNotFound() {
        txtTranslated.setText(R.string.translation_not_found);
    }

    @Override
    public void showTranslation(DisplayTranslateResult response) {
        txtTranslated.setText(response.textTranslated);
    }

    @Override
    public void showProgress() {
        progress.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideProgress() {
        progress.setVisibility(View.INVISIBLE);
    }

    @Override
    public void showErrorLayout() {
        errorLayout.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideErrorLayout() {
        errorLayout.setVisibility(View.INVISIBLE);
    }

    @Override
    public void setSrcLabel(String text) {
        srcLangBtn.setText(text);
    }

    @Override
    public void setDstLabel(String text) {
        dstLangBtn.setText(text);
    }

    @Override
    public void showOffline() {
        offlineIcon.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideOffline() {
        offlineIcon.setVisibility(View.INVISIBLE);
    }

    @Override
    public void showClearBtn() {
        clearBtn.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideClearBtn() {
        clearBtn.setVisibility(View.INVISIBLE);
    }

    @Override
    public void clearResult() {
        editInput.getText().clear();
    }

    @Override
    public void clearTranslation() {
        txtTranslated.setText("");
    }

    @Override
    public void onDestroy() {
        presenter.dispose();
        presenter = null;
        super.onDestroy();
    }
}

File: feature-translate/src/main/kotlin/com/andyanika/translator/feature/translate/TranslationFragment.kt
package com.andyanika.translator.feature.translate

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.lifecycle.lifecycleScope
import com.andyanika.translator.feature.translate.databinding.FragmentTranslationBinding
import com.andyanika.translator.feature.translate.mvi.TranslationAction
import com.andyanika.translator.feature.translate.mvi.TranslationFeature
import com.andyanika.translator.feature.translate.mvi.TranslationViewState
import com.andyanika.widgets.observe
import core.constants.Extras
import core.constants.Screens
import core.interfaces.ScreenRouter
import core.models.LanguageCode
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import org.koin.android.ext.android.getKoin
import org.koin.android.ext.android.inject
import org.koin.androidx.scope.ScopeFragment
import java.util.*

class TranslationFragment : ScopeFragment() {
    // NOTE: as a variant: manually link feature's scope to fragment's scope
    private val featureScope = getKoin()
        .getOrCreateScope<TranslationFeature>("")
        .apply { scope.linkTo(this) }

    private val feature by inject<TranslationFeature>()
    private val router: ScreenRouter by inject()

    private val binding by lazy { FragmentTranslationBinding.bind(requireView()) }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_translation, null)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        lifecycleScope.launchWhenStarted {
            feature
                .observe()
                .collect(::render)
        }

        lifecycleScope.launchWhenResumed {
            binding
                .editInput
                .observe()
                .collect {
                    feature.accept(TranslationAction.Translate(it))
                }
        }

        with(binding) {

            btnLangSrc.setOnClickListener {
                router.navigateTo(Screens.SELECT_LANGUAGE, Extras.MODE_SRC)
            }

            btnLangDst.setOnClickListener {
                router.navigateTo(Screens.SELECT_LANGUAGE, Extras.MODE_DST)
            }

            btnLangSwap.setOnClickListener {
                lifecycleScope.launch {
                    feature.accept(TranslationAction.SwapDirection)
                }
            }

            btnClear.setOnClickListener {
                lifecycleScope.launch {
                    feature.accept(TranslationAction.Clear)
                }
            }

            btnRetry.setOnClickListener {
                lifecycleScope.launch {
                    feature.accept(TranslationAction.Retry)
                }
            }
        }
    }

    override fun onStop() = with(binding) {
        // NOTE: as a variant: manually clear feature if not bound to fragment's lifecycle
        featureScope.close()

        super.onStop()
    }

    private fun showNotFound() {
        binding.txtTranslated.setText(R.string.translation_not_found)
    }

    private fun showTranslation(textTranslated: String) {
        binding.txtTranslated.text = textTranslated
    }

    private fun showProgress(visible: Boolean) {
        binding.searchProgress.isVisible = visible
    }

    private fun showError(visible: Boolean) {
        binding.errorLayout.isVisible = visible
    }

    private fun setSrcLabel(text: String?) {
        binding.btnLangSrc.text = text
    }

    private fun setDstLabel(text: String?) {
        binding.btnLangDst.text = text
    }

    private fun showOffline(visible: Boolean) {
        binding.iconOffline.isVisible = visible
    }

    private fun showClearBtn(visible: Boolean) {
        binding.btnClear.isVisible = visible
    }

    private fun render(state: TranslationViewState) = with(binding) {
        println("### $state")

        showProgress(state.isLoading)
        showError(state.isError)
        showOffline(state.isOffline)
        showClearBtn(state.canClear)
        if (state.isNotFound) showNotFound()

        setSrcLabel(getLanguageLabel(state.srcCode))
        setDstLabel(getLanguageLabel(state.dstCode))
        showTranslation(state.output)

        if (editInput.text.toString() != state.input) {
            editInput.text.replace(0, editInput.text.length, state.input)
        }
    }

    private fun getLanguageLabel(code: LanguageCode): String {
        val resourceName = "lang_" + code.toString().lowercase(Locale.getDefault())
        val id = resources.getIdentifier(resourceName, "string", requireActivity().packageName)
        return getString(id)
    }

    private var View.isVisible: Boolean
        get() {
            return visibility == View.VISIBLE
        }
        set(value) {
            visibility = if (value) View.VISIBLE else View.INVISIBLE
        }
}
